File: podklady/6_Others/2_github_copilot_sdk/python/1_single_agent/10_plugins/10_agent_with_plugins.py
"""
10 — Plugins.

Concept
-------
Copilot SDK doesn't ship a first-class "plugin" mechanism, but you can build
one trivially out of the primitives it *does* expose: skills + custom agents.
This file defines a tiny convention:

    plugin/
      plugin.json          # manifest: name, version, skills_dir, agents
      skills/<n>/SKILL.md  # discoverable skills
      ...

The `load_plugin()` helper below reads a manifest, returns the skill directory
and a list of custom-agent definitions, and we wire those into a session.

Run from the plugin folder:
    cd 1_single_agent/10_plugins
    python 10_agent_with_plugins.py
"""

import asyncio
import json
import os

from copilot import CopilotClient
from copilot.session import PermissionHandler


HERE = os.path.dirname(os.path.abspath(__file__))


def load_plugin(plugin_dir: str) -> tuple[str | None, list[dict]]:
    """Read a `plugin.json` and return (skill_dir, [custom_agent_dict, ...])."""
    with open(os.path.join(plugin_dir, "plugin.json")) as f:
        manifest = json.load(f)
    skills_dir = (
        os.path.join(plugin_dir, manifest["skills_dir"])
        if "skills_dir" in manifest
        else None
    )
    return skills_dir, manifest.get("agents", [])


async def main() -> None:
    client = CopilotClient()
    await client.start()

    skills_dir, agents = load_plugin(os.path.join(HERE, "demo-plugin"))

    session = await client.create_session(
        on_permission_request=PermissionHandler.approve_all,
        skill_directories=[skills_dir] if skills_dir else [],
        custom_agents=agents,
    )

    print("> Asking for an HTML page (the 'geocities-html' skill should kick in):")
    reply = await session.send_and_wait(
        "Generate an HTML page advertising a fictional 'Cyber Cat Café'."
    )
    print(reply.data.content if reply else "(no reply)")

    print("\n> Asking a generic question (the haiku-footer agent may take over):")
    reply = await session.send_and_wait(
        "Use the haiku-footer agent to answer: what is functional programming?"
    )
    print(reply.data.content if reply else "(no reply)")

    await session.disconnect()
    await client.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: podklady/6_Others/2_github_copilot_sdk/typescript/src/1_single_agent/10_plugins/10_agent_with_plugins.ts
/**
 * 10 — Plugins.
 *
 * Concept
 * -------
 * Copilot SDK doesn't ship a first-class "plugin" mechanism, but you can
 * build one trivially out of the primitives it *does* expose: skills + custom
 * agents. This file defines a tiny convention:
 *
 *   plugin/
 *     plugin.json         # manifest: name, version, skills_dir, agents
 *     skills/<n>/SKILL.md # discoverable skills
 *
 * The `loadPlugin()` helper below reads a manifest, returns the skill
 * directory and a list of custom-agent definitions.
 *
 * Run from this folder:
 *   cd src/1_single_agent/10_plugins
 *   npx tsx 10_agent_with_plugins.ts
 */

import { CopilotClient, approveAll } from "@github/copilot-sdk";
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));

interface PluginManifest {
    name: string;
    version: string;
    skills_dir?: string;
    agents?: Array<{
        name: string;
        displayName?: string;
        description?: string;
        tools?: string[];
        prompt: string;
    }>;
}

function loadPlugin(pluginDir: string): {
    skillsDir: string | null;
    agents: NonNullable<PluginManifest["agents"]>;
} {
    const manifest = JSON.parse(
        fs.readFileSync(path.join(pluginDir, "plugin.json"), "utf-8"),
    ) as PluginManifest;
    return {
        skillsDir: manifest.skills_dir ? path.join(pluginDir, manifest.skills_dir) : null,
        agents: manifest.agents ?? [],
    };
}

async function main(): Promise<void> {
    const client = new CopilotClient();

    const { skillsDir, agents } = loadPlugin(path.join(HERE, "demo-plugin"));

    const session = await client.createSession({
        onPermissionRequest: approveAll,
        skillDirectories: skillsDir ? [skillsDir] : [],
        customAgents: agents,
    });

    console.log("> Asking for an HTML page (the 'geocities-html' skill should kick in):");
    let reply = await session.sendAndWait({
        prompt: "Generate an HTML page advertising a fictional 'Cyber Cat Café'.",
    }, 300_000);
    console.log((reply?.data as { content: string })?.content ?? "(no reply)");

    console.log("\n> Asking a generic question (the haiku-footer agent may take over):");
    reply = await session.sendAndWait({
        prompt: "Use the haiku-footer agent to answer: what is functional programming?",
    }, 300_000);
    console.log((reply?.data as { content: string })?.content ?? "(no reply)");

    await session.disconnect();
    await client.stop();
}

main().catch(console.error);
